HW-3 MVP

juan-rybczinski opened the pull request:
- 다른 분들 로직 참고해 기본적인 기능 추가했습니다.
- MovieContract 정의해서 View, Presenter 구현했습니다.

juan-rybczinski committed:
HW-3 Implement basic functions before assignment
- Handling cases: No keywords, No results, Error occurred
- Hide keyboard when there is a search result
- Open url resource when each movie item is clicked

juan-rybczinski committed:
HW-3 Define MainContract

juan-rybczinski committed:
HW-3 Implement MoviePresenter based on MovieContract.Presenter
- Rename MainContract to MovieContract

juan-rybczinski committed:
HW-3 Convert MovieFragment to act as a View Component based on MovieContract.View

hunihun commented on class14/juan-rybczinski/GoogryAndroidArchitectureStudy/app/src/main/java/com/example/googryandroidarchitecturestudy/ui/fragment/MovieFragment.kt:
```diff
@@ -20,7 +25,11 @@ class MovieFragment : Fragment() {
     private val TAG = this::class.java.simpleName
 
-    private lateinit var movieAdapter: MovieAdapter
+    private val movieAdapter: MovieAdapter by lazy {
```
굳이 lazy 블록을 사용하지 않아도 될것같습니다!

juan-rybczinski replied:
movieAdapter가 최초 로딩 때부터 호출되니 굳이 lazy 초기화 할 필요가 없을 것 같아 제거했습니다. 감사합니다!

98e60a60c

hunihun replied:
:+1:

hunihun commented on class14/juan-rybczinski/GoogryAndroidArchitectureStudy/app/src/main/java/com/example/googryandroidarchitecturestudy/ui/fragment/MovieFragment.kt:
```diff
@@ -32,6 +41,10 @@ class MovieFragment : Fragment() {
         MovieRepositoryImpl(movieRemoteDataSource, movieLocalDataSource)
     }
 
+    private val inputMethodManager by lazy {
```
여기도 굳이 lazy 블록이 필요 없을듯 합니다!

juan-rybczinski replied:
키보드 제어는 현재 영화 검색 결과가 있을 경우에만 하고 있으니 굳이 최초 로딩 때 초기화 할 필요가 없을 것 같아 lazy 초기화 유지했습니다. 다만 불필요한 동기화로 인한 오버헤드를 줄이기 위해 `LazyThreadSafetyMode.NONE`으로 지정해 보았습니다.

b9b35d409

hunihun replied:
제가 잘못 이해하고 있었던것 같습니다!
👍👍👍

hunihun commented on class14/juan-rybczinski/GoogryAndroidArchitectureStudy/app/src/main/java/com/example/googryandroidarchitecturestudy/ui/fragment/MovieFragment.kt:
```diff
@@ -45,6 +48,10 @@ class MovieFragment : Fragment() {
         requireActivity().getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
     }
 
+    private val presenter: MovieContract.Presenter by lazy {
```
여기도 lazy 블록 사용이 불필요해보입니다!

juan-rybczinski replied:
movieAdapter와 마찬가지로 최초 로딩 때부터 presenter가 호출되고 있어 lazy 초기화 제거했습니다. 혹시 검색 버튼을 누를 때 setOnClickListener의 람다 함수 안의 presenter가 최초로 호출된다면 lazy 초기화를 유지하는 게 좋지 않을까하는 생각도 드네요.

c1253ee53

juan-rybczinski replied:
테스트 안해보고 푸시 했다가 에러가 나서 되돌렸습니다 😂
presenter를 멤버 변수로 초기화할 때 lazy 초기화 변수인 movieRepository를 참조했더니 에러가 나네요.
멘토님께서 수업 시간 때 말씀하셨던 내용이 이런 케이스 인가요?

6201d9343

hunihun replied:
버튼을 누를때라면 lazy 사용이 맞다고 생각합니다!
수업때 말씀하셨던건, lazy 블록안에 선언한 변수는 밖에서는 사용할수없던 케이스였던것 같네요!

sjjeong replied:
movieRepository안에서 MovieLocalDataSource를 만드는데 필요한 context 때문에 그렇습니다
context는 라이프싸이클과 관련이 있기 때문에 이런상황에서는 lazy를 써야해요

hunihun commented on class14/juan-rybczinski/GoogryAndroidArchitectureStudy/app/src/main/java/com/example/googryandroidarchitecturestudy/ui/presenter/MoviePresenter.kt:
```diff
@@ -23,7 +22,7 @@ class MoviePresenter(
             view.hideKeyboard()
```
로직이 필요없는 기능인것같은데 인터페이스로 사용하시는 이유가 있을까요 ??

juan-rybczinski replied:
키보드 제어도 View 컴포넌트의 동작이라고 생각해서 Interface에 정의해 보았습니다.

hunihun replied:
요기도 제가 잘못 이해하고 있는건가해서 석준님께 다시 여쭤보겠습니다!

hunihun reviewed (commented):
고생하셨습니다!

juan-rybczinski committed:
HW-3 Remove lazy init. of movieAdapter(https://github.com/StudyFork/GoogryAndroidArchitectureStudy/pull/901#discussion_r524340763)

juan-rybczinski committed:
HW-3 Set LazyThreadSafetyMode to None of lazy init. of inputMethodManager(https://github.com/StudyFork/GoogryAndroidArchitectureStudy/pull/901#discussion_r524341030)

juan-rybczinski committed:
HW-3 Remove lazy init. of presenter(https://github.com/StudyFork/GoogryAndroidArchitectureStudy/pull/901#discussion_r524342771)

juan-rybczinski committed:
HW-3 Roll back lazy init. of presenter(https://github.com/StudyFork/GoogryAndroidArchitectureStudy/pull/901#discussion_r524342771)

sjjeong reviewed (changes requested):
잘 하셨습니다~
Base하는 작업도 해보세요

juan-rybczinski committed:
HW-3 Create BaseContract to organize common functions
- Create contract package

juan-rybczinski committed:
HW-3 Implement BaseFragment and BasePresenter
- Handling fragment ViewBinding in BaseFragment
- Handling to hide keyboard in BaseFragment
- Make a polymorphic presenter in BaseFragment

juan-rybczinski committed:
HW-3 Rearrange inheritance
- MovieFragment and MoviePresenter inherit BaseFragment and BasePresenter respectively
- View and Presenter in MovieContract inherit View and Presenter in BaseContract

juan-rybczinski committed:
HW-3 Implement url validation check in BaseContract.Presenter when movie item is selected
- Make Movie domain inherit UrlResource interface
- Add functions: showUrlResource, showInvalidUrl in BaseContract.View

juan-rybczinski committed:
HW-3 Move common view logic from MovieContract to BaseContract

juan-rybczinski commented:
[jihoi-kang](https://github.com/StudyFork/GoogryAndroidArchitectureStudy/commit/a85698cc3aae7c3359d65daede2fe7eb1eba4e9f)님의 커밋을 많이 참고했는데(👍) 멘토님의 의도대로 제가 제대로 한 건지 잘 모르겠네요 ;

처음에 구조 변경이랑 리팩토링?을 동시에 하려니 엄청 헷갈렸는데 멘토님이 알려주신대로 껍데기 interface와 abstract class로 상속 구조 먼저 잡아두고 리팩토링 수행하면 훨씬 수월할 듯 하네요.

sjjeong reviewed (approved):
LGTM ✊ ✌️ 👋
